pkg/cmd/sponsors/list/list_test.go

```diff
@@ -0,0 +1,90 @@
+package list
+
+import (
+	"bytes"
+	"net/http"
+	"testing"
+
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/httpmock"
+	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewCmdList(t *testing.T) {
+	tests := []struct {
+		name  string
+		cli   string
+		wants string
+	}{
+		{
+			name:  "happy path",
+			cli:   "octocat",
+			wants: "octocat",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ios, _, _, _ := iostreams.Test()
+			f := &cmdutil.Factory{
+				IOStreams: ios,
+			}
+
+			argv, err := shlex.Split(tt.cli)
+			assert.NoError(t, err)
+
+			var gotOpts *ListOptions
+			cmd := NewCmdList(f, func(opts *ListOptions) error {
+				gotOpts = opts
+				return nil
+			})
+			cmd.SetArgs(argv)
+			cmd.SetIn(&bytes.Buffer{})
+			cmd.SetOut(&bytes.Buffer{})
+			cmd.SetErr(&bytes.Buffer{})
+
+			_, err = cmd.ExecuteC()
+			assert.NoError(t, err)
+
+			assert.Equal(t, tt.wants, gotOpts.Username)
+		})
+	}
+}
+
+func Test_listRun(t *testing.T) {
+	tests := []struct {
+		name  string
+		opts  *ListOptions
+		wants []string
+	}{
+		{
+			name: "happy path",
+			opts: &ListOptions{
+				HttpClient: func() (*http.Client, error) {
+					r := &httpmock.Registry{}
+
+					r.Register(
+						httpmock.GraphQL(`query SponsorsList\b`),
+						httpmock.StringResponse(`{"data": {"user": {"sponsors": {"totalCount": 2, "nodes": [{"login": "mona"}, {"login": "lisa"}]}}}}`))
+
+					return &http.Client{Transport: r}, nil
```

Comment:
🤔 I copied this pattern over from another test file and it gave me pause. This smells like over-testing to me... There's already graphql stubs in [api/queries_sponsors_test.go](api/queries_sponsors_test.go) - copy/pasted here, in fact. What my reflex was here was to look into how to mock out `api.GetSponsorsList()` in line 48 of [pkg/cmd/sponsors/list/list.go](pkg/cmd/sponsors/list/list.go):

> opts.Sponsors, err = api.GetSponsorsList(httpClient, "", opts.Username)

Here's my reasons:

1. It avoids duplicative GraphQL mocks
2. It reduces the scope of the test to the interfaces between the functions, making these tests orthogonal (as written, this test is technically also testing the functionality of `api.GetSponsorsList`, which feels weird)
3. It ensures that this test won't be impacted by changes to `api.GetSponsorsList` (we should be able to rely on type safety for interfaces, so the "contract" that creates protects them)

What are y'all's thoughts, @andyfeller and @williammartin?

Reply:
> I copied this pattern over from another test file and it gave me pause. This smells like over-testing to me...

I think touches on what Will will call an abstraction code smell as the logic responsible for providing the data might not be fake-able so tests like this have to resort to working at the HTTP layer.

Reply:
😬 Oh I have opinions on this alright.

Reply:
I think that your reflex is bang on. I believe that the previous maintainers tried to integrate most of each command together in tests to increase confidence. In my opinion however, it has come at the cost of maintening the test suite.

The tests become quite challenging to read with such significant set up in the table and the setup repeats itself at different layers. More so, for commands that _share_ API requests, instead of being clear and saying "we have something that plays that role, it has its own tests, and here we're just going to return a stubbed value" we have to repeat the API set up across every test.

I can talk more about this when I share my implementation. I don't want to eagerly share it though :)

Reply:
See my tinkering in [this PR](https://github.com/jtmcg/gh-cli/pull/1/files)